Allow a post to be marked as sticky

dnalves opened the pull request:
Fixes #7953

**This is still in progress, please don't merge**

### Description
We wanna allow the users to mark a post as sticky from the Android app. This is already present in iOS and Web.
A blocker for this PR is the approval of https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2124 and an update of the FluxC library (which will add the "sticky" field to the `PostModel`).

### Work done

A switch was added to Post Settings screen. This is loaded and updated in the `PostModel` (from FluxC), which syncs it to the Wordpress servers.
This switch only appears for Posts, so it is hidden for Pages.

### Screenshot
<img src="https://user-images.githubusercontent.com/628412/133897108-4f0a4a41-8911-4300-9e73-507567f71b8a.jpeg" width=350 />

### To be done

* Identify any regression areas
* Localize strings
* Add details to the release-notes

To test:
1. Open a site with posts in it
2. Open a post to edit it
3. Tap to edit "Post settings"
4. Toggle the switch "Stick post to the front page"
5. Update the post with the changes (sync it)
6. Check on Web or iOS that the changes were applied
7. Revert the change on Web or iOS
8. Check on Android that the change is reflected there

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests

3. What automated tests I added (or what prevented me from doing so)
There are no tests surrounding that Fragment. And any unit tests added will be testing the FluxC library.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dnalves committed:
Add a (dummy) Switch to toggle if the post is sticky in Post Settings

For now, it's still a dummy switch (due to FluxC needing to be updated).
We're updating its value when all the views are updated.

dnalves committed:
[PostSettings] Add logic to the "sticky" switch to update the PostModel

This depends on an update of the FluxC library (PR pending).

dnalves committed:
[PostSettings] Hide "sticky" switch when is a page (and not a post)

dnalves committed:
Update RELEASE-NOTES

Added "Allow users to mark Posts as sticky [https://github.com/wordpress-mobile/WordPress-Android/pull/15351]"

dnalves committed:
Update fluxCVersion to PR #2126

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -139,6 +143,9 @@
     @Inject ViewModelProvider.Factory mViewModelFactory;
     private EditPostPublishSettingsViewModel mPublishedViewModel;
 
+    private final OnCheckedChangeListener onStickySwitchChangeListener =
```
Checkstyle is suggesting that we adhere to the pattern of putting a `m` in front of the member variable. 
```suggestion
    private final OnCheckedChangeListener mOnStickySwitchChangeListener =
```

dnalves replied:
Oh, now I see that the Checkstyle job failed (but I can't see the details in BuildKite). I guess I should've ran the checks locally. Thanks for pointing it out :)

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -13,13 +13,16 @@
 import android.widget.ImageView.ScaleType;
 import android.widget.LinearLayout;
 import android.widget.TextView;
+import android.widget.Toast;
```
Since this import is not utilized what do you think about removing it?

dnalves replied:
Oh, this slipped 😭

jd-alexander commented on WordPress/src/main/res/layout/edit_post_settings_fragment.xml:
```diff
@@ -220,6 +220,30 @@
+            <LinearLayout style="@style/PostSettingsContainer">
+
+                <androidx.appcompat.widget.SwitchCompat
+                    android:id="@+id/post_settings_sticky_switch"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:text="@string/post_settings_post_sticky"
+                    android:textAlignment="viewStart"
+                    android:textAppearance="?attr/textAppearanceSubtitle1" />
```
What do you think about moving these attributes to `styles.xml` similarly to some of the other components here?

dnalves replied:
Make sense. My thought before was that it was used in a single place, so I decided to not extract them.

I just pushed a commit with them extracted to `styles.xml` 😄

dnalves committed:
[PostSettings] Rename Sticky switch listener and extract styles

dnalves committed:
Merge branch 'develop' into add-sticky-switch-to-post-settings

dnalves committed:
Update RELEASE-NOTES.txt

dnalves committed:
Update build.gradle to point fluxCVersion to the correct PR-commit

jd-alexander reviewed (commented):
The FluxC component of this PR has been merged in. A FluxC artifact is being built on FluxC's `develop` https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/e20192c7874769d07a9c6816adf9e047a81d954a

jd-alexander reviewed (changes requested):
@dnalves while testing and comparing both platforms I noticed that on iOS we are showing a `Sticky` label on the `Posts List Cell` as seen below in the screenshot. This was not included in the design specs that were attached to the issue. Nonetheless, in the name of platform synchronicity, we could make both `Posts Lists Item States` represent the same behavior when a post is sticky. What do you think?

<kbd><img src="https://user-images.githubusercontent.com/1509205/134222178-68a19c41-e88f-43cd-b6b6-130d130c2a58.png" width="320"></kbd>

dnalves committed:
Change FluxC version to the latest develop

jd-alexander commented:
> @dnalves while testing and comparing both platforms I noticed that on iOS we are showing a `Sticky` label on the `Posts List Cell` as seen below in the screenshot. This was not included in the design specs that were attached to the issue. Nonetheless, in the name of platform synchronicity, we could make both `Posts Lists Item States` represent the same behavior when a post is sticky. What do you think?
> 
> <img alt="" width="320" src="https://user-images.githubusercontent.com/1509205/134222178-68a19c41-e88f-43cd-b6b6-130d130c2a58.png">

Just a note here, from a design standpoint, if you agree with proceeding we could replicate this design approach i.e putting the label in a separate row within the `Posts List Item` as a first pass while we inquire about design feedback.

dnalves commented:
Nice catch @jd-alexander, it makes total sense for the behavior to be similar.

From what I noticed, the "Sticky" label is concatenated with the other labels at the bottom. We can replicate it. What do you think?

| Android  | iOS  | 
|---|---|
| <img src="https://user-images.githubusercontent.com/628412/134229902-e7b84ced-1fac-49cc-a0d5-b8582236ae16.jpeg" width=350 />  | <img src="https://user-images.githubusercontent.com/628412/134229911-f295c5b9-2af1-410b-beae-a29288b4b9c0.PNG" width=350 />  |

dnalves committed:
Display 'Sticky' label in the Post List for sticky posts

jd-alexander commented:
> Nice catch @jd-alexander, it makes total sense for the behavior to be similar.

I agree!

> From what I noticed, the "Sticky" label is concatenated with the other labels at the bottom. We can replicate it. What do you think?

Good observation! Sounds like an optimal approach that adheres to the current design specifications of the Posts List Item! I will check it out.

dnalves committed:
Fix detekt and klint for PostListItemUiStateHelper

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -3,6 +3,7 @@ package org.wordpress.android.viewmodel.posts
 import org.apache.commons.text.StringEscapeUtils
 import org.wordpress.android.BuildConfig
 import org.wordpress.android.R
+import org.wordpress.android.R.string
```
I am not sure if we should be making this change 🤔 I know it is used in other places within the codebase but it seems unrelated to the task at hand here. If it's a case where the CI is failing due to it then I guess it's fine to proceed as it might be a new rule I am not aware of. Let me know your thoughts.

dnalves replied:
Android Studio function extraction did this and I didn't notice. When I read your comment I was surprised by it as well. I'm fixing this!

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -233,44 +234,69 @@ class PostListItemUiStateHelper @Inject constructor(
         }
+    }
+
+    @Suppress("LongParameterList")
```
I know this solution works. I have seen it in usage elsewhere but I know we are not fond of doing suppressions. Part of me feels it would be better if we simply pass the `PostModel` here and create these values within the function to lessen the parameters substantially. Do you have any thoughts on this discourse?

dnalves replied:
I'm not a fan as well, to be honest. I added here because I assumed in this case the long parameter list wasn't making the function more complex or harder to understand (given that the parameters were well named).

But I like your solution 😄  I'll pass the `PostModel` instead of a few parameters.

jd-alexander replied:
Thanks! I remember one of the concerns for passing around the`PostModel` is that it's quite a big object that contains so much information about the post that we may not require. To ensure we are making the best decision here, I am going to loop in another Android developer who has worked with these kind of decisions. 

Hi @ParaskP7 I remember us making a decision about `@Suppress("LongParameterList")` in one of my PRs a few months back. In this case, we were opting to remove it and pass in a `PostModel` to represent several parameters that were extracted from it. However, we know the `PostModel` is quite a large object. Do you have any thoughts on this? Let me know. Thank you.

ParaskP7 replied:
👋 @jd-alexander @dnalves !

> I remember one of the concerns for passing around thePostModel is that it's quite a big object that contains so much information about the post that we may not require.
> In this case, we were opting to remove it and pass in a PostModel to represent several parameters that were extracted from it. However, we know the PostModel is quite a large object.

Thanks for looping me in to this conversation Joel. I would personally go with passing the whole `PostModel` and thus resolving the `LongParameterList` warning. My reasoning would be:
1. The `PostModel` is already in memory, so I am not sure if it being a large object would degrade performance or cause any trouble. I haven't looked at the code extensively, but are you having concerns about passing this object between screens, thus having is unnecessarily being serialized and deserialised, or is it just being passed from one class/method to another?
2. I am not sure how big could potentially the `PostModel` end up being, thus I might be totally off here, but I would not try to optimize performance preemptively by sending individual properties of this object. However, this is my personal rule of thump, which might very well not apply here, so please take it with a grain of salt.

Let me know if I can further help here or whether I am missing some more context. 🙏

dnalves replied:
👋 @ParaskP7 

Thanks for the quick response on this matter 😄 I agree with your points. About your question:

> [...] are you having concerns about passing this object between screens, thus having is unnecessarily being serialized and deserialised, or is it just being passed from one class/method to another?

This is just from one method to another, so no serialization occurring in this case.

@jd-alexander: I made those changes in https://github.com/wordpress-mobile/WordPress-Android/pull/15351/commits/982bd0fad599a8e2d1e376adf861151d233d9654, already resolving the `LongParameterList` 😄

jd-alexander replied:
> The PostModel is already in memory, so I am not sure if it being a large object would degrade performance or cause any trouble. I haven't looked at the code extensively, but are you having concerns about passing this object between screens, thus having is unnecessarily being serialized and deserialised, or is it just being passed from one class/method to another?

It's just being passed to another method and yes it is already in memory :) 

> I am not sure how big could potentially the PostModel end up being, thus I might be totally off here, but I would not try to optimize performance preemptively by sending individual properties of this object. However, this is my personal rule of thumb, which might very well not apply here, so please take it with a grain of salt.

> Let me know if I can further help here or whether I am missing some more context. 🙏

The thoughts you shared were spot on. You picked up the context quite quickly, just like an Android API using `requireContext()` 😄 😆 

 Thanks for chiming in @ParaskP7 . I think the current approach here is fine!

dnalves committed:
A few code improvements in PostListItemUiStateHelper

Change which import we're using for R.string.
Remove a detekt supression and fix the LongParameterList issue

jd-alexander commented:
@dnalves I was doing some more testing and I noticed something. 

1. If you create a post and you mark it as `Sticky` the list item displays `Sticky` in black color. 
2. If you create a post and you mark it as `Sticky` and `Private`, the list item displays it in yellow color. If you remove the `Private` status, it causes an inconsistency because the `Sticky` label still remains yellow and others look black. What do you think about making it yellow by default? I think all labels have a color on Android that is not black. 

<kbd><img src="https://user-images.githubusercontent.com/1509205/134567592-769954cd-28ca-4977-9cf7-40c07d95f67f.png" width="320"></kbd>

jd-alexander commented:
@dnalves since there were some new changes in FluxC we can use the latest `develop` hash before merging these changes. The changes made there are related to WooCommerce, so WP Android would be fine. Normally, we can merge both PRs when both are approved but it's also fine that it was merged in before since it doesn't cause any breaking changes if it were FluxC were to be updated in WP Android with another PR. Just noting that here :)

jd-alexander commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -253,28 +250,53 @@ class PostListItemUiStateHelper @Inject constructor(
+        hasUnhandledConflicts: Boolean,
+        hasAutoSave: Boolean
+    ): List<UiString> {
+        val labels: MutableList<UiString> = getErrorAndProgressStatuses(
```
Isn't the type of labels inferred by the compiler? Let me know what you think.

dnalves replied:
Oh, nice catch! Android Studio method extraction tool made this and I missed it, thanks for pointing it out :D 
https://github.com/wordpress-mobile/WordPress-Android/pull/15351/commits/985ed1f301283ab2986551cdbf11b84021b91521#diff-2238863dfadc135bc1b6359a21a855541d2dcf3d8cd2742d2fa91a81ea03e701L288

dnalves committed:
Refactor PostPageListLabelColorUseCase to use the Post instead of multiple params

dnalves committed:
Change color of the sticky label to same as other info labels (yellow)

dnalves commented:
> @dnalves I was doing some more testing and I noticed something.
> 
> 1. If you create a post and you mark it as `Sticky` the list item displays `Sticky` in black color.
> 2. If you create a post and you mark it as `Sticky` and `Private`, the list item displays it in yellow color. If you remove the `Private` status, it causes an inconsistency because the `Sticky` label still remains yellow and others look black. What do you think about making it yellow by default? I think all labels have a color on Android that is not black.
> 

You're totally right! I missed this because I was testing only with private posts (which makes all the labels yellow). I added the "sticky" label to the list of "info" labels, which are colored yellow.

https://github.com/wordpress-mobile/WordPress-Android/pull/15351/commits/9f8175b96fd7f4ed0ba4f84c05b9ac1171065cd0

dnalves committed:
Update FluxC version to the latest develop commit

dnalves committed:
Merge remote-tracking branch 'upstream/develop' into add-sticky-switch-to-post-settings

jd-alexander commented:
> > @dnalves I was doing some more testing and I noticed something.
> > 
> > 1. If you create a post and you mark it as `Sticky` the list item displays `Sticky` in black color.
> > 2. If you create a post and you mark it as `Sticky` and `Private`, the list item displays it in yellow color. If you remove the `Private` status, it causes an inconsistency because the `Sticky` label still remains yellow and others look black. What do you think about making it yellow by default? I think all labels have a color on Android that is not black.
> 
> You're totally right! I missed this because I was testing only with private posts (which makes all the labels yellow). I added the "sticky" label to the list of "info" labels, which are colored yellow.
> 
> [9f8175b](https://github.com/wordpress-mobile/WordPress-Android/commit/9f8175b96fd7f4ed0ba4f84c05b9ac1171065cd0)


Thanks for these changes @dnalves Could you please update the [tests](https://github.com/dnalves/WordPress-Android/blob/c3867f0e062baf1e90424f1de4013037c0f4d7aa/WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PostPageListLabelColorUseCaseTest.kt#L15) to represent this behavior as well. When modifying files in WP Android, it's always best to check if there are associated tests ( which you have done previously 😎 , so I am just noting here as a general reminder.) 👍🏾 


I normally Command-click a class to verify if tests are associated or not. 

![tests](https://user-images.githubusercontent.com/1509205/134719646-a672187d-7b2c-453f-8bc9-db7f3b9a83c1.png)

dnalves committed:
Add unit tests for Sticky label color use case

dnalves commented:
> > > @dnalves I was doing some more testing and I noticed something.
> > > 
> > > 1. If you create a post and you mark it as `Sticky` the list item displays `Sticky` in black color.
> > > 2. If you create a post and you mark it as `Sticky` and `Private`, the list item displays it in yellow color. If you remove the `Private` status, it causes an inconsistency because the `Sticky` label still remains yellow and others look black. What do you think about making it yellow by default? I think all labels have a color on Android that is not black.
> > 
> > 
> > You're totally right! I missed this because I was testing only with private posts (which makes all the labels yellow). I added the "sticky" label to the list of "info" labels, which are colored yellow.
> > [9f8175b](https://github.com/wordpress-mobile/WordPress-Android/commit/9f8175b96fd7f4ed0ba4f84c05b9ac1171065cd0)
> 
> Thanks for these changes @dnalves Could you please update the [tests](https://github.com/dnalves/WordPress-Android/blob/c3867f0e062baf1e90424f1de4013037c0f4d7aa/WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PostPageListLabelColorUseCaseTest.kt#L15) to represent this behavior as well. When modifying files in WP Android, it's always best to check if there are associated tests ( which you have done previously 😎 , so I am just noting here as a general reminder.) 👍🏾
> 
> I normally Command-click a class to verify if tests are associated or not.
> 
> ![tests](https://user-images.githubusercontent.com/1509205/134719646-a672187d-7b2c-453f-8bc9-db7f3b9a83c1.png)

😱 
Thanks for letting me know, my bad.
I added the tests here: https://github.com/wordpress-mobile/WordPress-Android/pull/15351/commits/ff5191b6917f1a4768a772d1d80bb64bcbeec19d

dnalves committed:
Merge remote-tracking branch 'upstream/develop' into add-sticky-switch-to-post-settings

dnalves committed:
Update FluxC version to the latest develop

jd-alexander committed:
Merge branch 'develop' into add-sticky-switch-to-post-settings

jd-alexander reviewed (approved):
Thanks for these changes @dnalves LGTM 🚢
